doc/neps/nep-0031-uarray.rst

```diff
@@ -0,0 +1,364 @@
+not to force the conversion. ``ua.Dispatchable`` is a simple class consisting
+of three simple values: ``type``, ``value``, and ``coercible``.
+``__ua_convert__`` returns an iterable of the converted values, or
+``NotImplemented`` in the case of failure. Returning ``NotImplemented``
+here will cause ``uarray`` to move to the next available backend.
+
+``__ua_function__`` has the signature ``(func, args, kwargs)`` and defines
+the actual implementation of the function. It recieves the function and its
+arguments. Returning ``NotImplemented`` will cause a move to the default
+implementation of the function if one exists, and failing that, the next
+backend.
+
+If all backends are exhausted, a ``ua.BackendNotImplementedError`` is raised.
+
+Backends can be registered for permanent use if required.
+
+Defining overridable methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To define an overridable method, one needs a few things:
+
+1. A dispatcher that returns an iterable of ``ua.Dispatchable`` objects.
+2. A reverse dispatcher that replaces dispatchable values with the supplied
+   ones.
+3. A domain.
+4. Optionally, a default implementation, which can be provided in terms of
+   other overridable methods.
+
+As an example, consider the following::
+
+    import uarray as ua
+
+    def full_argreplacer(args, kwargs, dispatchables):
+        def full(shape, fill_value, dtype=None, order='C'):
+            return (shape, fill_value), dict(
+                dtype=dispatchables[0],
+                order=order
+            )
+
+        return full(*args, **kwargs)
+
+    @ua.create_multimethod(full_argreplacer, domain="numpy")
+    def full(shape, fill_value, dtype=None, order='C'):
+        return (ua.Dispatchable(dtype, np.dtype),)
+
+A large set of examples can be found in the ``unumpy`` repository, `[8]`_.
+This simple act of overriding callables allows us to override:
+
+* Methods
+* Properties, via ``fget`` and ``fset``
+* Entire objects, via ``__get__``.
+
+Using overrides
+~~~~~~~~~~~~~~~
+
+The way we propose the overrides will be used by end users is::
+
+    import numpy.overridable as np
+    with np.set_backend(backend):
+        x = np.asarray(my_array, dtype=dtype)
+
+And a library that implements a NumPy-like API will use it in the following
+manner (as an example)::
+
+    import numpy.overridable as np
+    _ua_implementations = {}
+
+    __ua_domain__ = "numpy"
+
+    def __ua_function__(func, args, kwargs):
+        fn = _ua_implementations.get(func, None)
+        return fn(*args, **kwargs) if fn is not None else NotImplemented
+
+    def implements(ua_func):
+        def inner(func):
+            _ua_implementations[ua_func] = func
+            return func
+
+        return inner
+
+    @implements(np.asarray)
+    def asarray(a, dtype=None, order=None):
+        # Code here
+        # Either this method or __ua_convert__ must
+        # return NotImplemented for unsupported types,
+        # Or they shouldn't be marked as dispatchable.
```

Comment:
Does this imply that the ``__array__`` magic method should be deprecated?

Reply:
Not at all, that one is for returning a NumPy array, this is for duck arrays.

Reply:
> Should `__ua_convert__` be defined here, or can it be undefined reasonably? Which I guess signals that any input is expected to be fine?

Yes, the absence of `__ua_convert__` signals that anything is fine, and no conversion takes place. As does the absence of a reverse dispatcher/argument replacer (which, to be clear, are two names for the same thing).

> As to returning NotImplemented. Is there a chance of that choking up things because some function/method may actually want to return NotImplemented for a reason? (I guess that is unlikely).

I think it's standard practice in Python for protocols to do something special when `NotImplemented` is returned, and it is only ever meant to be returned from a protocol.